properties.go

```diff
@@ -0,0 +1,65 @@
+package smithy
+
+// PropertiesReader provides an interface for reading metadata from the
+// underlying metadata container.
+type PropertiesReader interface {
+	Get(key interface{}) interface{}
+}
+
+// Properties provides storing and reading metadata values. Keys may be any
+// comparable value type. Get and set will panic if key is not a comparable
+// value type.
+//
+// Properties uses lazy initialization, and Set method must be called as an
+// addressable value, or pointer. Not doing so may cause key/value pair to not
+// be set.
+type Properties struct {
+	values map[interface{}]interface{}
```

Comment:
We may want to consider strongly typing these (I know the SEP says not to, but there's nothing wrong with doing so if we explicitly mark the types as not stable IMO), this would simplify consuming it in runtime down the road (you'd just pull a concrete instance of something out of the property bag rather than having to construct it from what is essentially a string -> void* map on the fly).

That's what we did in kotlin, and we actually also have an extensibility mechanism for a white-label SDK implementor to generate their own typed values.

I don't know that we necessarily _must_ to go this way in go, you could still do well with compartmentalized abstractions to deal with pulling out the raw map at runtime, but we should at least be aware of the option.

Reply:
i see, did you have a particular type in mind? say, `map[string]string`?

Reply:
I think this is fine it mirrors context API and we can't introduce generics anywhere so this looks ok to me for Go.

Reply:
discussed offline with @lucix-aws @aajtodd . leaving this for now, and will prob add in strong typing at the accessor method level at the end